gapis/api/gles/dependencygraph2_test.go

```diff
@@ -0,0 +1,189 @@
+	if cmdNode, ok := node.(dependencygraph2.CmdNode); ok {
+		cmdId := cmdNode.Index[0]
+		if api.CmdID(cmdId).IsReal() {
+			command := graph.GetCommand(api.CmdID(cmdId))
+			return command.CmdName()
+		}
+	}
+	return ""
+}
+
+// testCaptureAndGraph abstracts the boilerplate for creating a capture
+// programmatically, check expected dependencies in its graph, and
+// clean out
+type testCaptureAndGraph struct {
+	ctx   context.Context
+	arena arena.Arena
+	cb    gles.CommandBuilder
+	cmds  []api.Cmd
+}
+
+// init runs the boilerplate up to being able to add commands to tc.cmds
+func (tc *testCaptureAndGraph) init(t *testing.T) {
+	tc.ctx = log.Testing(t)
+	tc.ctx = bind.PutRegistry(tc.ctx, bind.NewRegistry())
+	tc.ctx = database.Put(tc.ctx, database.NewInMemory(tc.ctx))
+	tc.arena = arena.New()
+	ctxHandle := memory.BytePtr(1)
+	displayHandle := memory.BytePtr(2)
+	surfaceHandle := memory.BytePtr(3)
+	tc.cb = gles.CommandBuilder{Thread: 0, Arena: tc.arena}
+
+	// Common prologue: make an EGL context
+	tc.cmds = []api.Cmd{
+		tc.cb.EglCreateContext(displayHandle, surfaceHandle, surfaceHandle, memory.Nullptr, ctxHandle),
+		api.WithExtras(
+			tc.cb.EglMakeCurrent(displayHandle, surfaceHandle, surfaceHandle, ctxHandle, 0),
+			gles.NewStaticContextStateForTest(tc.arena), gles.NewDynamicContextStateForTest(tc.arena, 64, 64, false)),
+	}
+}
+
+// terminate cleans up what need to be
+func (tc *testCaptureAndGraph) terminate() {
+	tc.arena.Dispose()
+}
+
+// checkDependenciesArePresent generate the graph and check if
+// expected dependencies present
+func (tc *testCaptureAndGraph) checkDependenciesArePresent(t *testing.T, toCheck []testDependencyCmdName) {
+	// Compute dependency graph
+	header := &capture.Header{ABI: device.AndroidARM64v8a}
+	cap, err := capture.NewGraphicsCapture(tc.ctx, tc.arena, t.Name(), header, nil, tc.cmds)
+	if err != nil {
+		panic(err)
+	}
+	capturePath, err := cap.Path(tc.ctx)
+	if err != nil {
+		panic(err)
+	}
+	tc.ctx = capture.Put(tc.ctx, capturePath)
+
+	cfg := dependencygraph2.DependencyGraphConfig{
+		MergeSubCmdNodes:       true,
+		IncludeInitialCommands: false,
+	}
+	graph, err := dependencygraph2.GetDependencyGraph(tc.ctx, capturePath, cfg)
+	if err != nil {
+		panic(err)
+	}
+
+	// Check dependencies
+	mapToCheck := map[testDependencyCmdName]bool{}
+	for _, dep := range toCheck {
+		mapToCheck[dep] = false
+	}
+
+	graph.ForeachDependency(
+		func(src, tgt dependencygraph2.NodeID) error {
+			srcCmdName := getCommandName(graph, src)
+			tgtCmdName := getCommandName(graph, tgt)
+			dep := testDependencyCmdName{srcCmdName, tgtCmdName}
+			if _, ok := mapToCheck[dep]; ok {
+				mapToCheck[dep] = true
+			}
+
+			return nil
+		})
+
+	for dep, present := range mapToCheck {
+		assert.For(tc.ctx, fmt.Sprintf("Dependency: %v", dep)).ThatBoolean(present).IsTrue()
+	}
+}
+
+// Actual tests
+
+// glClear(GL_COLOR_BUFFER_BIT) depends on glClearColor()
+func TestDependencyGlClearColor(t *testing.T) {
```

Comment:
One comment I would like to make is that a dependency graph that just returned "KEEP EVERYTHING" would also pass these tests. 

If we want to write unit tests for the dependency graph, we should also at least make sure that some of the edges are NOT present.

e.g. 
glClear(GL_COLOR_BUFFER_BIT) does not depend on glClearDepthf()

Obviously we can't be exhaustive, but we should at least attempt in some cases to make sure it is not a complete graph.

Reply:
Very good point, I updated the unit test template to take two list of dependencies: expected and unexpected dependencies.